### PR TITLE
Tips

### PR DESCRIPTION
**Pull request by StudentAlleg**
Framework for adding a chat Tip system.

Currently sends a message to players not in a Clan that they can join a Clan. 

Plans to expand include added a toggle in settings (partially setup), other tip information.

### CODE REVIEW

File: clans/src/main/java/me/mykindos/betterpvp/clans/clans/tips/TipEvent.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import net.minecraft.world.entity.player.Player;
+
+@Getter
+public class TipEvent extends CustomCancellableEvent {
+
+    private final Gamer gamer;
+    TipEvent(Gamer gamer){
+        super(true);
+        this.gamer = gamer;
+    }
```
Comment (Mykindos) [inline clans/src/main/java/me/mykindos/betterpvp/clans/clans/tips/TipEvent.java:17]: You dont need this if you put @Data annotation on the TipEvent

File: clans/src/main/java/me/mykindos/betterpvp/clans/clans/tips/TipListener.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        Gamer gamer;
+        if (gamerOptional.isPresent()) {
+            gamer = gamerOptional.get();
+            gamer.setLastTip(System.currentTimeMillis());
+        }
+    }
+
+    @UpdateEvent(delay = 10 * 1000, isAsync = true)
+    public void tipSender() {
+        gamerManager.getObjectsValues().forEach(gamer -> {
```
Comment (Mykindos) [inline clans/src/main/java/me/mykindos/betterpvp/clans/clans/tips/TipListener.java:51]: for(Player player : Bukkit.getOnlinePlayers()) {
    gamerManager.getObject(player.getUniqueId()).ifPresent(gamer -> {
        // tip logic
    });
}

File: core/src/main/java/me/mykindos/betterpvp/core/gamer/GamerManager.java [hunk 3]
```diff
@@ -31,6 +32,10 @@ public Optional<Gamer> getGamerByName(String name) {
         return objects.values().stream().filter(gamer -> gamer.getClient().getName().equalsIgnoreCase(name)).findFirst();
     }
 
+    public Collection<Gamer> getObjectsValues() {
```
Comment (Mykindos) [inline core/src/main/java/me/mykindos/betterpvp/core/gamer/GamerManager.java:35]: Collection<?> getObjectsValues() 

in Manager.java

**Review by Mykindos — CHANGES_REQUESTED** [review]

File: clans/src/main/java/me/mykindos/betterpvp/clans/clans/tips/TipListener.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            if (gamer.getPlayer() != null){
+                UtilServer.runTaskLaterAsync(clans, () -> UtilServer.callEvent(new TipEvent(gamer)), 5);
+            }
+        });
+
+    }
+
+    @EventHandler
```
Comment (Mykindos) [inline clans/src/main/java/me/mykindos/betterpvp/clans/clans/tips/TipListener.java:59]: @EventHandler (priority = EventPriority.MONITOR)

Also check at the start of the method if event.isCancelled()

**Review by Mykindos — APPROVED** [review]

**Comment by StudentAlleg** [comment]
Updated

**Comment by StudentAlleg** [comment]
Updated framework to use Enums. This is kind of how I want to do things, but I am looking to refactor into a weighted list format to allow for semi-random choice of what tip to send.

**Comment by StudentAlleg** [comment]
Rebased in preparation for review.

**Review by StudentAlleg — COMMENTED** [review]
Ready for Review. Changes: Tips are now stored in Enums (is this too much). Implemented WighedList via TipList for easily handling inserting tips. Added hover and click events.

**Comment by StudentAlleg** [comment]
tips

**Comment by Mykindos** [comment]
It looks ok, though I'm not sure if enum is the right choice. You could go down the road of creating a seperate object for each tip and give each of them their own predicate. That way in the tip event you wouldn't need to check a bunch of different conditions manually, would make adding new tips a bit easier

**Review by StudentAlleg — COMMENTED** [review]
Changed Tips to be implemented similarly to how skills are. There are probably improvements like predicates, but I am not completely sure how to use them. 

It should be much easier to add new skills.

**Comment by StudentAlleg** [comment]
Also cleaned up and removed a lot of added files that turned out to be unnecessary with this implementation

**Review by StudentAlleg — COMMENTED** [review]
Moving Tips to core per feedback.

**Review by StudentAlleg — COMMENTED** [review]
Tried changing to a loader style, didn't work. https://github.com/StudentAlleg/BetterPvP/tree/student/tip-loader

Works as is.

**Review by Mykindos — APPROVED** [review]

**Pull request merged**